image/descriptor.go

```diff
@@ -48,19 +50,36 @@ func (d *descriptor) hash() string {
 	return pts[1]
 }
 
+// Index implementation
+type Index v1.ImageIndex
+
 func listReferences(w walker) (map[string]*descriptor, error) {
 	refs := make(map[string]*descriptor)
+	var index Index
 
 	if err := w.walk(func(path string, info os.FileInfo, r io.Reader) error {
-		if info.IsDir() || !strings.HasPrefix(path, "refs") {
+		if info.IsDir() || filepath.Clean(path) != "index.json" {
 			return nil
 		}
 
-		var d descriptor
-		if err := json.NewDecoder(r).Decode(&d); err != nil {
+		buf, err := ioutil.ReadAll(r)
```

Comment:
doing a `.ReadAll` is a bit more expensive, where the previous version is a streaming decoder. I'd prefer a streaming decoder.